text/0000-no-entry.md

```diff
@@ -0,0 +1,149 @@
+have limitations on such support. Targets that do not support this should
+document the target-specific limitations of `no_entry` for their target, may
+require `no_std` in addition to `no_entry`, or may not support `no_entry` at
+all.
+
+On targets that already don't provide an entry point, the compiler will
+silently accept `no_entry`. On targets that cannot support `no_entry`, the
+compiler will emit a compile-time error.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Rust has several existing attributes related to program startup, and this would
+require further careful documentation to distinguish them. However, no existing
+attribute serves the function that `no_entry` does.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Today, people do this either by linking their binary manually, or by specifying
+non-portable linker arguments in `.cargo/config`. Introducing the `no_entry`
+attribute substantially simplifies creating such programs, and avoids
+target-specific build steps.
+
+The name `no_entry` was selected specifically to minimize conflict and
+confusion with other Rust features such as `start`, `no_main`, and `main`, none
+of which relate to the entry-point symbol.
+
+We could potentially integrate this mechanism with specification of an
+alternate entry point. However, this would make it difficult to flexibly
+provide the entry point from a library or other mechanism.
+
+We could make this a rustc command-line option, rather than an attribute.
+However, that would separate the configuration required to build a crate from
+the crate itself. We have substantial precedent for including such
+configuration directly in the crate as an attribute.
+
+# Prior art
+[prior-art]: #prior-art
+
+We currently have the `no_main` attribute to omit a C `main` function (not the
+entry point symbol), the `main` attribute to specify a *Rust* `main` function
+other than `fn main()`, and the `start` attribute to specify a Rust-specific
+portable startup function that runs after some existing Rust startup code.
+
+We do not currently have any mechanisms to address handling of entry-point
+symbols.
+
+Existing programming languages rely entirely on the toolchain for this
+functionality. Introducing `no_entry` makes it possible to gain control from
+the initial binary entry point without toolchain-specific code, consistent with
+other Rust efforts to provide some uniformity across targets and toolchains.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Can we use common code to implement this for all targets using a given
+toolchain family (`LinkerFlavor`), and minimize target-specific implementation
+code?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+We should simplify the declaration of an entry-point symbol, such as by
+providing an `#[entry]` attribute. The symbol itself would still have a
```

Comment:
If we have `#[entry]` on `unsafe ... fn foobar(...)` do we then also need `#![no_entry]`?

Reply:
FWIW this is an elaboration of https://github.com/rust-lang/rfcs/pull/2735/files/e4ad78dd444961bbc268f4f4a46189999750a5ca#r308998642

Reply:
@Centril Yes, we do. `entry` could be used in a library, for instance, with a binary crate using `no_entry` and using that library. I'll add a note that it would make sense to have a lint if you use `entry` in a binary crate without using `no_entry`, though.

Reply:
@joshtriplett Ah I see; so this is akin to "needs / provides" of `#![requires_allocator]` (because every program requires an entry point) and `#[global_allocator]`. -- Please mention this similarity explicitly as well as the use-case for library/binary division in the text also. :)

Reply:
What do you mean by `requires_allocator`? I don't see any references to that or similar names anywhere on the web.

The RFC already mentions the use case of library/binary division, in the rationale and alternatives section: "However, this would make it difficult to flexibly provide the entry point from a library or other mechanism."

Reply:
> What do you mean by `requires_allocator`? I don't see any references to that or similar names anywhere on the web.

Ooops; It's called `#![needs_allocator]` in https://github.com/rust-lang/rust/blob/master/src/liballoc/lib.rs#L65 apparently.

> The RFC already mentions the use case of library/binary division, in the rationale and alternatives section: "However, this would make it difficult to flexibly provide the entry point from a library or other mechanism."

Yes, but connecting this to `entry` and `no_entry` is another matter, it would be helpful to clarify that this is what is referred to by that sentence.